src/TestProblem/Hydro/Zeldovich/Init_TestProb_Zeldovich.cpp

```diff
@@ -0,0 +1,537 @@
+static int NPar_YZ;              // 1D number of particles along the y/z-axis
+static double Pert_Wave_Len;     // perturbation wavelength [Mpc/h]
+static double k_Pert;            // perturbation wavenumber
+static double zi_Initial;        // redshfit of simulation IC
+static double zc_Collapse;       // redshfit at the onset of Zeldovich pancake collapse
+static double GasTemp_Init;      // initial gas temperature in Kelvin
+static double H0;                // Hubble parameter at redshift zero
+static double rho_crit_i;        // comoving critical density = 3H0^2/8*pi*G
+static double dhx;               // 1D inter-particle separation [Mpc/h]
+// =======================================================================================
+
+
+// problem-specific function prototypes
+struct Zeldovich_coord_params
+{
+   double x_shift_input, zi_Initial;
+};
+static double Zeldovich_coord_transformation(double x_Lagrangian_Trial, void *params);
+static void OutputError();
+#ifdef SUPPORT_GSL
+   static double Zeldovich_coord_transformation_solver(double x_shift_input, double zi_Initial);
+   static double Zeldovich_density_profile(double x_Lagrangian, double zi_Initial);
+   static double Zeldovich_x_velocity_profile(double x_Lagrangian, double zi_Initial);
+#endif
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Validate
+// Description :  Validate the compilation flags and runtime parameters for this test problem
+//
+// Note        :  None
+//
+// Parameter   :  None
+//
+// Return      :  None
+//-------------------------------------------------------------------------------------------------------
+void Validate()
+{
+
+   if ( MPI_Rank == 0 )    Aux_Message( stdout, "   Validating test problem %d ...\n", TESTPROB_ID );
+
+
+// errors
+#  if ( MODEL != HYDRO )
+   Aux_Error( ERROR_INFO, "MODEL != HYDRO !!\n" );
+#  endif
+
+#  ifndef GRAVITY
+   Aux_Error( ERROR_INFO, "GRAVITY must be enabled !!\n" );
+#  endif
+
+#  ifndef COMOVING
+   Aux_Error( ERROR_INFO, "COMOVING must be enabled !!\n" );
+#  endif
+
+#  ifdef COMOVING
+   if ( OMEGA_M0 != 1 )
+      Aux_Error( ERROR_INFO, "OMEGA_M0 != 1 (matter-only) for this test !!\n");
+#  endif
+
+#  ifdef GRAVITY
+   if ( OPT__BC_FLU[0] != BC_FLU_PERIODIC  ||  OPT__BC_POT != BC_POT_PERIODIC )
+      Aux_Error( ERROR_INFO, "must adopt periodic BC for this test !!\n" );
+#  endif
+
+   if ( OPT__INIT != INIT_BY_FUNCTION  &&  OPT__INIT != INIT_BY_RESTART )
+      Aux_Error( ERROR_INFO, "OPT__INIT != FUNCTION (1) or RESTART (2) for this test !!\n" );
+
+#  ifndef FLOAT8
+   Aux_Error( ERROR_INFO, "FLOAT8 must be enabled !!\n" );
+#  endif
+
+#  ifndef SUPPORT_GSL
+   Aux_Error( ERROR_INFO, "SUPPORT_GSL must be enabled !!\n" );
+#  endif
+
+#  ifdef PARTICLE
+   if ( amr->Par->Init != PAR_INIT_BY_FUNCTION )
+      Aux_Error( ERROR_INFO, "PAR_INIT != FUNCTION (1) for this test !!\n" );
```

Comment:
Add `  &&  amr->Par->Init != PAR_INIT_BY_RESTART`

Reply:
Do you mean changing line 95 to `if ( amr->Par->Init != PAR_INIT_BY_FUNCTION && amr->Par->Init != PAR_INIT_BY_RESTART )` ? If so, I don't quite understand the modification here since we want to enforce the option `amr->Par->Init == PAR_INIT_BY_FUNCTION` in this test problem, right? Or if you are concerned about particle initialization during RESTART, I guess the following modification would be more appropriate `if ( amr->Par->Init != PAR_INIT_BY_FUNCTION || amr->Par->Init != PAR_INIT_BY_RESTART ) ... Aux_Error( ... )`.

Reply:
I'm indeed considering particle initialization during restart. Note that `PAR_INIT` will be reset to `2` automatically when adopting `OPT__INIT = 2` ([here](https://github.com/gamer-project/gamer/blob/main/src/Init/Init_ResetParameter.cpp#L815-L824)). But I still think the check should be `&&` instead of `||`.